### ISSUE TITLE
官方demo可以提供吗

### ISSUE THREAD

**Issue by weibiaoyi** [Kwai-Keye/Keye#37]

**Comment by najwagober9003-collab** [comment]
`import os
import mimetypes
from typing import List, Dict, Any, Tuple, Optional

import gradio as gr
import torch
from transformers import AutoModel, AutoProcessor
from keye_vl_utils import process_vision_info


# ------------------------------
# Lazy loader for model & processor
# ------------------------------
_MODEL = None
_PROCESSOR = None


def get_model_and_processor():
    global _MODEL, _PROCESSOR
    if _MODEL is not None and _PROCESSOR is not None:
        return _MODEL, _PROCESSOR

    model_path = os.getenv("KEYE_MODEL", "Kwai-Keye/Keye-VL-1_5-8B")

    _MODEL = AutoModel.from_pretrained(
        model_path,
        torch_dtype="auto",
        device_map="auto",
        trust_remote_code=True,
    )
    _PROCESSOR = AutoProcessor.from_pretrained(model_path, trust_remote_code=True)
    return _MODEL, _PROCESSOR


# ------------------------------
# Helpers
# ------------------------------
def _detect_file_type(path: str) -> Optional[str]:
    if not path:
        return None
    mime, _ = mimetypes.guess_type(path)
    if not mime:
        return None
    if mime.startswith("image/"):
        return "image"
    if mime.startswith("video/"):
        return "video"
    return None


def _append_user_message(
    mm_messages: List[Dict[str, Any]], user_text: str, file_path: Optional[str]
) -> List[Dict[str, Any]]:
    content: List[Dict[str, Any]] = []
    if file_path:
        kind = _detect_file_type(file_path)
        if kind == "image":
            content.append({"type": "image", "image": file_path})
        elif kind == "video":
            content.append({"type": "video", "video": file_path})
    if user_text and user_text.strip():
        content.append({"type": "text", "text": user_text.strip()})
    if not content:
        return mm_messages
    mm_messages = list(mm_messages)
    mm_messages.append({"role": "user", "content": content})
    return mm_messages


@torch.inference_mode()
def generate_reply(mm_messages: List[Dict[str, Any]], max_new_tokens: int = 512) -> str:
    model, processor = get_model_and_processor()

    text = processor.apply_chat_template(
        mm_messages, tokenize=False, add_generation_prompt=True
    )
    image_inputs, video_inputs, mm_kwargs = process_vision_info(mm_messages)

    inputs = processor(
        text=[text],
        images=image_inputs,
        videos=video_inputs,
        padding=True,
        return_tensors="pt",
        **mm_kwargs,
    )

    # Move tensors to CUDA if available. Model uses device_map="auto".
    device = "cuda" if torch.cuda.is_available() else "cpu"
    inputs = inputs.to(device)

    generated_ids = model.generate(**inputs, max_new_tokens=max_new_tokens)
    generated_ids_trimmed = [
        out_ids[len(in_ids) :] for in_ids, out_ids in zip(inputs.input_ids, generated_ids)
    ]
    output_texts = processor.batch_decode(
        generated_ids_trimmed, skip_special_tokens=True, clean_up_tokenization_spaces=False
    )
    return output_texts[0] if output_texts else ""


# ------------------------------
# Gradio handlers
# ------------------------------
def on_submit(
    user_text: str,
    upload_file,
    chat_history: List[Tuple[str, str]],
    mm_messages: List[Dict[str, Any]],
    think_mode: str,
    max_tokens: int,
):
    file_path = None
    if isinstance(upload_file, str):
        file_path = upload_file
    elif upload_file is not None and hasattr(upload_file, "name"):
        file_path = upload_file.name
    # Apply thinking suffix per README (
    display_text = user_text or ""
    text_for_model = display_text
    if (think_mode or "").lower() == "think":
        text_for_model = (display_text.strip() + "/think") if display_text else "/think"
    elif (think_mode or "").lower() in ("no think", "no_think", "no-think"):
        text_for_model = (display_text.strip() + "/no_think") if display_text else "/no_think"

    mm_messages = _append_user_message(mm_messages, text_for_model, file_path)
    if not mm_messages or mm_messages[-1]["role"] != "user":
        return chat_history, mm_messages, gr.update(value=""), None

    reply = generate_reply(mm_messages, max_new_tokens=int(max_tokens or 512))
    mm_messages.append({"role": "assistant", "content": [{"type": "text", "text": reply}]})

    chat_history = list(chat_history)
    chat_history.append((display_text or "", reply))
    return chat_history, mm_messages, gr.update(value=""), None


def on_regenerate(
    chat_history: List[Tuple[str, str]],
    mm_messages: List[Dict[str, Any]],
    max_tokens: int,
):
    if not chat_history or not mm_messages:
        return chat_history, mm_messages

    # Remove last assistant message if exists
    if mm_messages and mm_messages[-1]["role"] == "assistant":
        mm_messages = list(mm_messages[:-1])

    reply = generate_reply(mm_messages, max_new_tokens=int(max_tokens or 512))
    mm_messages.append({"role": "assistant", "content": [{"type": "text", "text": reply}]})

    # Update last tuple's assistant side
    chat_history = list(chat_history)
    user_last, _ = chat_history[-1]
    chat_history[-1] = (user_last, reply)
    return chat_history, mm_messages


def on_clear():
    return [], [] , gr.update(value=""), None


def build_ui() -> gr.Blocks:
    with gr.Blocks(title="Kwai Keye-VL-1.5-8B Demo") as demo:
        gr.Markdown("**Kwai Keye-VL-1.5-8B Demo** — image/video understanding with multi-turn chat.")

        chat = gr.Chatbot(height=420, layout="panel", type="tuples")
        state_mm = gr.State([])  # multimodal messages in Keye format

        with gr.Row():
            txt = gr.Textbox(label="Input", lines=2, placeholder="Type your question...")

        with gr.Row():
            uploader = gr.File(label="Upload(上传文件)", file_types=["image", "video"], interactive=True)

        with gr.Row():
            think = gr.Radio(
                choices=["Auto", "Think", "No Think"],
                value="Auto",
                label="Thinking Mode",
            )
            max_tok = gr.Slider(
                minimum=64,
                maximum=2048,
                step=64,
                value=512,
                label="Max new tokens",
            )

        with gr.Row():
            btn_submit = gr.Button("Submit(发送)")
            btn_regen = gr.Button("Regenerate(重试)")
            btn_clear = gr.Button("Clear History(清除历史)")

        # Events
        btn_submit.click(
            fn=on_submit,
            inputs=[txt, uploader, chat, state_mm, think, max_tok],
            outputs=[chat, state_mm, txt, uploader],
        )

        btn_regen.click(
            fn=on_regenerate,
            inputs=[chat, state_mm, max_tok],
            outputs=[chat, state_mm],
        )

        btn_clear.click(
            fn=on_clear,
            inputs=None,
            outputs=[chat, state_mm, txt, uploader],
        )

        # Warmup toggle to lazily load when user first interacts, but allow manual preload if desired
        with gr.Accordion("Advanced", open=False):
            gr.Markdown("Model will load on first use. Set `KEYE_MODEL` env var to override.")

    return demo


if __name__ == "__main__":
    demo = build_ui()
    demo.queue(max_size=64).launch(server_name="0.0.0.0", server_port=int(os.getenv("PORT", "7860")))


`
I created a gradio app, but I'm not sure why its effect is worse than the official one
